File: py/20.有效的括号.py
#
# @lc app=leetcode.cn id=20 lang=python3
#
# [20] 有效的括号
#

# @lc code=start
class Solution:
    # 1，用栈存放左括号
    #  * 2，用map存放合法的括号对，键存右括号，值存左括号
    #  * 3，扫描字符串，当扫描到左括号是入栈，扫描到右括号是判断栈顶是否是合肥的左括号
    # 注意stack = list()比stack=[]快
    # 32ms    97%    76%   13.3MB
    def isValid1(self, s: str) -> bool:
        stack = list()
        parentheses_map = {')':'(',']':'[','}':'{'}
        for c in s:
            if parentheses_map.get(c, None) is  None:
                stack.append(c)
            elif len(stack)==0:
                return False
            elif parentheses_map[c] != stack.pop() :
                return False
        return len(stack) == 0
        # 跟方法1一样，只是判断条件不一样
    def isValid2(self, s: str) -> bool:
        stack = list()
        parentheses_map = {')':'(',']':'[','}':'{'}
        for c in s:
            if c in parentheses_map:
                if len(stack) == 0:
                    return False
                if parentheses_map[c] != stack.pop():
                    return False
            else:
                stack.append(c)
        return len(stack) == 0
    #  * 1，栈结构不变
    #  * 2，使用判断语句代替map
    #  * 此例判断语句写的很繁琐，不容易看懂
    # 40ms  50%   60%   13.4MB
    def isValid(self, s: str) -> bool:
        stack = list()
        for c in s:
            if c == '(':
                stack.append(')')
            elif c == '[':
                stack.append(']')
            elif c == '{':
                stack.append('}')
            elif len(stack) == 0 or stack.pop() != c:
                return False
        return len(stack) == 0

# @lc code=end

File: java/20.有效的括号.java
/*
 * @lc app=leetcode.cn id=20 lang=java
 *
 * [20] 有效的括号
 */

// @lc code=start
class Solution {
    /**1，用栈存放左括号
     * 2，用map存放合法的括号对，键存右括号，值存左括号
     * 3，扫描字符串，当扫描到左括号是入栈，扫描到右括号是判断栈顶是否是合肥的左括号
     * 3ms     26%    15%   37.1MB
     */
    public boolean isValid1(String s) {
        Deque<Character> stack = new LinkedList<>();
        Map<Character,Character> map = new HashMap<>(){{
            put(')','(');
            put(']','[');
            put('}','{');
        }
        };
        for(int i=0;i<s.length();i++){
            char tmp = s.charAt(i);
            if(map.containsKey(tmp)){
                if(stack.isEmpty()||stack.peek()!=map.get(tmp)){
                    return false;
                }
                stack.pop();
            }else{
                stack.push(tmp);
            }
        }
        if(stack.isEmpty())return true;
        return false;

    }
    /**
     * 1，栈结构不变
     * 2，使用判断语句代替map
     * 此例判断语句写的很繁琐，不容易看懂
     * 2ms   78%   27%   37MB
     */
    public boolean isValid2(String s) {
        Deque<Character> stack = new LinkedList<>();
        for(int i=0;i<s.length();i++){
            char tmp = s.charAt(i);
            if((tmp == '('||tmp == '['||tmp == '{')){
                stack.push(tmp);
                continue;
            }
            if(stack.isEmpty() && (tmp == ')'||tmp == ']'||tmp == '}') )
            return false;

            if((tmp == ')'&&stack.peek()=='(')||(tmp == '}'&&stack.peek()=='{')||(tmp == ']'&&stack.peek()=='[')){
                stack.pop();
            }else{
                return false;
            }
        }
        if(stack.isEmpty()) return true;
        return false;
    }
    /**
     * 与算法二逻辑一样，只是if语句写的更精简，更易懂
     * 2ms 78%  30%    37MB
     */
    public boolean isValid(String s) {
        Deque<Character> stack = new LinkedList<>();
        for(int i=0;i<s.length();i++){
            char tmp = s.charAt(i);
            if(tmp == '(' )stack.push(')');
            else if(tmp == '[')stack.push(']');
            else if(tmp == '{')stack.push('}');
            else if(stack.isEmpty() || tmp != stack.pop() ) return false;
        }
        return stack.isEmpty();

    }
}
// @lc code=end

